### PR TITLE
[Feature] MaxCompute(ODPS) Catalog FE Metadata parts

### PR DESCRIPTION
**Pull request by dingxin-tech**
Why I'm doing:
The integration between StarRocks and MaxCompute (originally known as ODPS) is crucial for users who need to access their data stored in MaxCompute without the hassle of data migration or duplication. To enhance our users' experience and provide more flexibility in data management, I'm introducing the ODPS Catalog feature to StarRocks.

What I'm doing:
I'm adding a new feature to StarRocks that allows users to connect to ODPS data sources directly. This feature, named ODPS Catalog, acts as a bridge between StarRocks and MaxCompute, giving users the ability to perform queries on MaxCompute datasets from within the StarRocks environment. This integration not only simplifies the data workflow but also ensures that users can leverage the full power of StarRocks's analytics capabilities on their existing data in MaxCompute.

Fixes #issue
#32719

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/OdpsPartitionKey.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.List;
+
+public class OdpsPartitionKey extends PartitionKey implements NullablePartitionKey {
+    public OdpsPartitionKey() {
+        super();
+    }
+
+    @Override
+    public List<String> nullPartitionValueList() {
+        return ImmutableList.of(JDBCTable.PARTITION_NULL_VALUE);
```
[the remark this answers is not in the document]
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/catalog/OdpsPartitionKey.java:29]:
```suggestion
        return ImmutableList.of(OdpsTable.PARTITION_NULL_VALUE);
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/OdpsTable.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        this.tableName = odpsTable.getName();
+        this.partitionColumns =
+                odpsTable.getSchema().getPartitionColumns().stream().map(EntityConvertUtils::convertColumn).collect(
+                        Collectors.toList());
+        this.dataColumns = fullSchema;
+        this.fullSchema = new ArrayList<>(dataColumns);
+        this.fullSchema.addAll(partitionColumns);
+    }
+
+    public String getProjectName() {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/catalog/OdpsTable.java:78]: I think just using getDbName as project name in the sr.

File: fe/fe-core/src/main/java/com/starrocks/catalog/Table.java [hunk 3]
```diff
@@ -594,6 +602,8 @@ public String getEngine() {
             return "JDBC";
         } else if (this instanceof FileTable) {
             return "File";
+        } else if (this instanceof OdpsTable) {
+            return "Odps";
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Table.java:606]: delete it

Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/catalog/Table.java]: accept

File: fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                catalogOwner = js.get("DisplayName").getAsString();
+            }
+            Iterator<Project> iterator = odps.projects().iterator(catalogOwner);
+            while (iterator.hasNext()) {
+                Project project = iterator.next();
+                builder.add(project.getName());
+            }
+        } catch (OdpsException odpsException) {
+            odpsException.printStackTrace();
+            throw new StarRocksConnectorException("fail to list project names", odpsException);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java:149]:
```suggestion
            throw new StarRocksConnectorException("fail to list project names", e);
```

File: fe/fe-core/src/main/java/com/starrocks/catalog/OdpsPartitionKey.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/catalog/OdpsPartitionKey.java]: yep，only ‘null'.

File: fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        // TODO: perhaps not good to support users to fetch whole tables?
+        return get(partitionCache, odpsTableName).stream()
+                .map(p -> p.toString(false, true)).collect(
+                        Collectors.toList());
+    }
+
+    @Override
+    public List<String> listPartitionNamesByValue(String databaseName, String tableName,
+                                                  List<Optional<String>> partitionValues) {
+        List<PartitionSpec> partitionSpecs = get(partitionCache, OdpsTableName.of(databaseName, tableName));
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java]: Currently, it is not available.

File: fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+                JsonObject js = JsonParser.parseString(result).getAsJsonObject();
+                catalogOwner = js.get("DisplayName").getAsString();
+            }
+            Iterator<Project> iterator = odps.projects().iterator(catalogOwner);
+            while (iterator.hasNext()) {
+                Project project = iterator.next();
+                builder.add(project.getName());
+            }
+        } catch (OdpsException odpsException) {
+            throw new StarRocksConnectorException("fail to list project names", odpsException);
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java:149]: accept

File: fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                odps.tables().get(odpsTableName.getDatabaseName(), odpsTableName.getTableName());
+        List<Partition> partitions = odpsTable.getPartitions();
+        return partitions.stream().map(Partition::getPartitionSpec).collect(Collectors.toList());
+    }
+
+    @Override
+    public List<PartitionInfo> getPartitions(Table table, List<String> partitionNames) {
+        OdpsTable odpsTable = (OdpsTable) table;
+        ImmutableList.Builder<PartitionInfo> builder = ImmutableList.builder();
+        odps.tables().get(odpsTable.getProjectName(), odpsTable.getTableName()).getPartitions()
```
[the remark this answers is not in the document]
Comment (zdx19981006) [reply fe/fe-core/src/main/java/com/starrocks/connector/odps/OdpsMetadata.java]: yes, I've fixed now.

**Review by stephen-shelby — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Comment by zdx19981006** [comment]
Hello,

Some of the commits in this pull request were not signed, causing them to fail the DCO check. I have gone ahead and resubmitted a new pull request with all the necessary signed commits.

You can find the new PR at the following link:

https://github.com/StarRocks/starrocks/pull/37423

Thank you for addressing this issue.

Best regards

**Review by stephen-shelby — DISMISSED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**